codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java

```diff
@@ -0,0 +1,526 @@
+        goAwayFrame.setExtraStreamIds(2);
+
+        inboundHandler.writeOutbound(releaseLater(goAwayFrame));
+        verify(frameWriter).writeGoAway(
+                eq(http2HandlerCtx), eq(7), eq(Http2Error.NO_ERROR.code()), eq(debugData), anyChannelPromise());
+        assertEquals(1, debugData.refCnt());
+        assertEquals(State.OPEN, stream.state());
+        assertTrue(channel.isActive());
+    }
+
+    @Test
+    public void receiveGoaway() throws Exception {
+        frameListener.onGoAwayRead(http2HandlerCtx, 2, Http2Error.NO_ERROR.code(), bb("foo"));
+
+        Http2GoAwayFrame expectedFrame = new DefaultHttp2GoAwayFrame(2, Http2Error.NO_ERROR.code(), bb("foo"));
+        Http2GoAwayFrame actualFrame = inboundHandler.readInbound();
+
+        assertEquals(releaseLater(expectedFrame), releaseLater(actualFrame));
+        assertNull(inboundHandler.readInbound());
+    }
+
+    @Test
+    public void unknownFrameTypeShouldThrowAndBeReleased() throws Exception {
+        class UnknownHttp2Frame extends AbstractReferenceCounted implements Http2Frame {
+            @Override
+            public String name() {
+                return "UNKNOWN";
+            }
+
+            @Override
+            protected void deallocate() {
+            }
+
+            @Override
+            public ReferenceCounted touch(Object hint) {
+                return this;
+            }
+        }
+
+        UnknownHttp2Frame frame = new UnknownHttp2Frame();
+        assertEquals(1, frame.refCnt());
+
+        ChannelFuture f = channel.write(frame);
+        f.await();
+        assertTrue(f.isDone());
+        assertFalse(f.isSuccess());
+        assertThat(f.cause(), instanceOf(UnsupportedMessageTypeException.class));
+        assertEquals(0, frame.refCnt());
+    }
+
+    @Test
+    public void incomingStreamActiveShouldFireUserEvent() throws Exception {
+        frameListener.onHeadersRead(http2HandlerCtx, 3, request, 31, false);
+
+        Http2Stream stream = framingCodec.connectionHandler().connection().stream(3);
+        assertNotNull(stream);
+
+        Http2HeadersFrame frame = inboundHandler.readInbound();
+        assertNotNull(frame);
+
+        Http2StreamActiveEvent streamActiveEvent = inboundHandler.readUserEvent();
+        assertEquals(stream.id(), streamActiveEvent.streamId());
+
+        assertNull(inboundHandler.readInbound());
+        assertNull(inboundHandler.readUserEvent());
+    }
+
+    @Test
+    public void goAwayLastStreamIdOverflowed() throws Exception {
+        frameListener.onHeadersRead(http2HandlerCtx, 5, request, 31, false);
+
+        Http2Stream stream = framingCodec.connectionHandler().connection().stream(5);
+        assertNotNull(stream);
+        assertEquals(State.OPEN, stream.state());
+
+        ByteBuf debugData = bb("debug");
+        Http2GoAwayFrame goAwayFrame = new DefaultHttp2GoAwayFrame(Http2Error.NO_ERROR.code(), debugData);
```

Comment:
do we need to `slice()` this buffer so that the equality comparison below doesn't fail?